lib/cleanse.rb

```diff
@@ -70,31 +66,41 @@ def self.same_fill_units(capacities, previous, current)
     end
   end
 
+  def self.useful_unit_ids(capacity_unit_ids, stop)
+    stop.loads.map{ |loa| loa.quantity.unit.id }.compact & capacity_unit_ids
+  end
+
   def self.cleanse_empties_fills(vrp, solution)
     service_types = %i[pickup delivery service]
     solution.routes.each{ |route|
       vehicle = route.vehicle
-      capacities_units = vehicle.capacities.collect{ |capacity| capacity.unit_id if capacity.limit }.compact
+      capacity_unit_ids = vehicle.capacities.collect{ |capacity| capacity.unit_id if capacity.limit }.compact
       previous_activity = nil
 
       route.stops.delete_if{ |stop|
         next unless service_types.include?(stop.type)
 
         if previous_activity && stop && same_position(vrp, stop) &&
-           same_empty_units(capacities_units, previous_activity, stop) &&
-           !same_fill_units(capacities_units, previous_activity, stop)
+           same_empty_units(capacity_unit_ids, previous_activity, stop) &&
+           !same_fill_units(capacity_unit_ids, previous_activity, stop)
           add_unnassigned(solution.unassigned_stops, stop, 'Duplicate empty service.')
           true
         elsif previous_activity && stop && same_position(vrp, stop) &&
-              same_fill_units(capacities_units, previous_activity, stop) &&
-              !same_empty_units(capacities_units, previous_activity, stop)
+              same_fill_units(capacity_unit_ids, previous_activity, stop) &&
+              !same_empty_units(capacity_unit_ids, previous_activity, stop)
           add_unnassigned(solution.unassigned_stops, stop, 'Duplicate fill service.')
           true
         else
           previous_activity = stop if previous_activity.nil? || stop.service_id
           false
         end
       }
+
+      empty_index = get_empty_index_if_first_position(capacity_unit_ids, route.stops)
+      if empty_index
+        removed_stop = route.remove_stop(vrp, route.stops[empty_index], empty_index)
```

Comment:
unused function argument
```suggestion
        removed_stop = route.remove_stop(vrp, empty_index)
```